lower/llpcSpirvLowerGlobal.cpp

```diff
@@ -1407,13 +1408,112 @@ Value* SpirvLowerGlobal::AddCallInstForInOutImport(
                     inOutInfo.SetInterpLoc(interpLoc);
                     inOutInfo.SetInterpMode(inOutMeta.InterpMode);
                 }
-                pInOutValue = m_pBuilder->CreateReadGenericInput(pInOutTy,
-                                                                 inOutMeta.Value,
-                                                                 pLocOffset,
-                                                                 pElemIdx,
-                                                                 maxLocOffset,
-                                                                 inOutInfo,
-                                                                 pVertexIdx);
+
```

Comment:
Please move this scalarizing logic into the middle-end, probably into PatchResourceCollect along with the rest of the logic. Also, once everything is localized better in PatchResourceCollect, you can move the logic of whether the optimization is enabled out of PipelineContext (which will soon be a front-end-only object) and into PatchResourceCollect too. Thanks.

Reply:
I don't think it is a good idea at that time.
- until now, LLPC builder doesn't include the real implementation of in/out import/export, i.e. we can't do the io packing in LLPC Builder.
- so It looks strange to me if LLPC builder only do the scalarization  but do io packing in patch pass.
- so I prefer keep current solution in the first commit, and continue to refine it if we have a better idea during implement the next phase.

Reply:
I am not suggesting putting the scalarization into Builder. I am suggesting putting it into PatchResourceCollect along with the packing logic. It can still be a separate step within that pass.

Reply:
The problem with putting the scalarization into a lower pass is that (a) it implicitly changes the Builder interface for inputs and output to only accept scalars for no good reason, and (b) even worse, it does that under control of a global option, so sometimes vectors are ok. That is not a clean interface.

Reply:
Move it to PatchResourceCollect  is OK to me. but I don't think it must be done immediately, and nothing is improved with that change, maybe we have to add an additional loop in  PatchResourceCollect. We have hold this change too long time, especially, this change affects some resource usage, it may conflict with your huge change. so I prefer to do it in next change, after your huge change is merged in.

Reply:
Ok, agreed, let's get this change in, and then fix the front/middle-end split thing in a later change.

Reply:
I have marked this conversation as resolved, but I don't know how to stop it saying that I am requesting changes...

Reply:
You can approve the change once again. Then the blocker mark will disappear.

Reply:
Aha, thanks.